Add Spacechem as an humblebundle game

layus opened the pull request.

aszlig commented on pkgs/games/humblebundle/spacechem.nix:
```diff
@@ -0,0 +1,69 @@
+  in ''
+    set -x
+    mkdir $out
+    ar p $src data.tar.gz | (cd $out; tar xvzf - ./opt)
```
Maybe something like `unpackCmd = "ar p \"$curSrc\" data.tar.gz | tar xz opt";`?

layus replied:
Well, this unpacks directly to $out, avoiding an extra move/copy step. is there a way to mimic that with unpackCmd ?

aszlig replied:
You could also specify `-C "$out"` then.

As for the size issue... I'm quite aware that the whole `unpack -> build -> install` is something that really could be improved. I don't know about Spacechem, but something like eg. Pyre has > 6 GB to unpack.

So the long-term solution would probably be something like unzipping the game's package directly into a store path (as a fixed-output derivation) and reference them from within the `buildGame` derivation. However, this could require more patching and would mess with the directory structure (also it wouldn't work so well with the md5 hashes for HumbleBundle, but I guess it's time to simply deprecate them anyway, regardless of whether HB will move on or not). Another idea I had is to use the sandbox to overlay paths directly via bind mounts into the chroot.

layus commented:
@aszlig I fixed according to your comments.

I had troubles with {setS,s}ourceRoot and fixed some code in buildGame. I am also not too sure about coreutils and uname invocations made by the game.

If this is ok, do not forget to autosquash before merging.

layus commented:
@aszlig Wanna have a second look at this one ?

aszlig commented on pkgs/games/build-support/build-game.nix:
```diff
@@ -30,7 +29,7 @@ buildSandbox (stdenv.mkDerivation ({
   setSourceRoot = ''
     popd &> /dev/null
-  '' + lib.optionalString (setSourceRoot == "") ''
+  '' + attrs.setSourceRoot or ''
```
What's the reason for this change to be included in the main commit adding the game?

layus replied:
None. It was part of another change and somehow landed here.

aszlig replied:
This is still not resolved, could you please remove this change from the commit?

layus replied:
Oh, crap ;-)

layus replied:
Done !

layus commented:
Updated, as a new fixed version was released in the meantime.

layus commented:
@aszlig Could you have a look at #18, #25 and #21 ?

layus commented:
Oh, BTW, the humblebundle downloader does not work anymore. Not sure if this is really important, especially given that the upstream python lib has not been updated since 2015.

aszlig commented:
@layus: Hm, one question remains though: You haven't set a sandbox path, so I'd assume the game writes save files in `$XDG_DATA_HOME` or `$XDG_CONFIG_HOME`, right? And if yes, it might be a good idea, to set a more specific path.

aszlig commented:
@layus: Regarding the HB downloader, I'm already aware of that and will fix it along with the merge of your PRs.

layus commented:
Updated the sandbox path. There is a strange access to a pulse "cookie" in CONFIG_HOME/pulse. It gets created by spacechem, but it needs write permission there.

aszlig commented:
@layus: While this is just an ugly thing that's not a stopper for merging this PR, can you post a `strace -f` of the game somewhere so we can figure out why this is is needed?

layus commented:
Here it is https://gist.github.com/dab849ee90e7384b45616c07379064e0 :-).

I grep'd 'layus' for brevity.

layus commented:
and the same, when "$XDG_CONFIG_HOME/pulse" is removed from the required sandbox paths: https://gist.github.com/3fb6dcb1de51a351d3ee16f49894a5aa

layus commented:
https://www.reddit.com/r/spacechem/comments/1qundh/fully_functional_spacechem_on_linux/ has interesting remarks. I have no idea what xclip is for. never encountered the bug.

aszlig commented on pkgs/games/humblebundle/default.nix:
```diff
@@ -28,6 +28,7 @@ let
     owlboy = callPackage ./owlboy.nix {};
     pico-8 = callPackage ./pico-8.nix {};
     rocketbirds = callPackage ./rocketbirds.nix {};
+    spacechem = callPackage ./spacechem.nix {};
```
As mentioned in [NixOS/nixpkgs#66620](https://github.com/NixOS/nixpkgs/pull/66620#pullrequestreview-294718277), it seems that the game is i686-only, so why not use `callPackage_i686`?

```suggestion
    spacechem = callPackage_i686 ./spacechem.nix {};
```

aszlig commented:
@layus: Regarding the pulse cookie, this might be also due to a mixture of 32/64 bit libraries as well, because usually the pulse cookie should be gathered via the X session, but for some reason not included in the strace (as it's only `-e trace=%file`), it wasn't able to do that.

aszlig commented:
> https://www.reddit.com/r/spacechem/comments/1qundh/fully_functional_spacechem_on_linux/ has interesting remarks. I have no idea what xclip is for. never encountered the bug.

I guess Spacechem is calling [`xclip`](https://github.com/NixOS/nixpkgs/blob/aeca9e6dcedb7839a93d22fa93807c3c9bceda0e/pkgs/tools/misc/xclip/default.nix) to put data into the user's clipboard. If that's the case, there is [`monogamePatcher`](https://github.com/openlab-aux/vuizvui/tree/cfaa888f6c38c325336b14f8cc6775cf7da95388/pkgs/games/build-support/monogame-patcher), which can be used to patch the assemblies. It's just a guess and it could also be Mono itself, so I'd double-check with `monodis` and look for clipboard references.

layus commented:
It is a .exe (admitedly, ad 32bit .exe) that runs with mono. It works perfectly well with a 64 bit mono, so I do not get the point of downloading all the 32 bit version of all the dependencies just to autoPatchelfHook patch rgb2theora correctly.

layus commented:
It looks like it is working properly with 32bits.

layus commented:
I could not find the source that calls xclip, but adding it to path fixes the issue. It's good enough for now.

Also removed the "cookie" sandbox path. Seems not needed for sound to play.
